storage/innobase/include/clone0desc.h

```diff
@@ -0,0 +1,650 @@
+@return true, if valid locator */
+bool clone_validate_locator(const byte *desc_loc, uint desc_len);
+
+/** Clone descriptors contain meta information needed for applying cloned data.
+These are PODs with interface to serialize and deserialize them. */
+enum Clone_Desc_Type {
+  /** Logical pointer to identify a clone operation */
+  CLONE_DESC_LOCATOR = 1,
+
+  /** Metadata for a Task/Thread for clone operation */
+  CLONE_DESC_TASK_METADATA,
+
+  /** Information for snapshot state */
+  CLONE_DESC_STATE,
+
+  /** Metadata for a database file */
+  CLONE_DESC_FILE_METADATA,
+
+  /** Information for a data block */
+  CLONE_DESC_DATA,
+
+  /** Must be the last member */
+  CLONE_DESC_MAX
+};
+
+/** Header common to all descriptors. */
+struct Clone_Desc_Header {
+  /** Descriptor version */
+  uint m_version;
+
+  /** Serialized length of descriptor in bytes */
+  uint m_length;
+
+  /** Descriptor type */
+  Clone_Desc_Type m_type;
+
+  /** Serialize the descriptor header: Caller must allocate
+  the serialized buffer.
+  @param[out]   desc_hdr        serialized header */
+  void serialize(byte *desc_hdr);
+
+  /** Deserialize the descriptor header.
+  @param[in]    desc_hdr        serialized header
+  @param[in]    desc_len        descriptor length
+  @return true, if successful. */
+  bool deserialize(const byte *desc_hdr, uint desc_len);
+};
+
+/** Task information in clone operation. */
+struct Clone_Task_Meta {
+  /** Index in task array. */
+  uint m_task_index;
+
+  /** Current chunk number reserved by the task. */
+  uint m_chunk_num;
+
+  /** Current block number that is already transferred. */
+  uint m_block_num;
+};
+
+/** Map for current block number for unfinished chunks. Used during
+restart from incomplete clone operation. */
+using Chunk_Map = std::map<uint32_t, uint32_t>;
+
+/** Bitmap for completed chunks in current state */
+class Chnunk_Bitmap {
+ public:
+  /** Construct bitmap */
+  Chnunk_Bitmap() : m_bitmap(), m_size(), m_bits() {}
+
+  /** Bitmap array index operator implementation */
+  class Bitmap_Operator_Impl {
+   public:
+    /** Construct bitmap operator
+    @param[in]  bitmap  reference to bitmap buffer
+    @param[in]  index   array operation index */
+    Bitmap_Operator_Impl(uint32_t *&bitmap, uint32_t index)
+
+        : m_bitmap_ref(bitmap) {
+      /* BYTE position */
+      auto byte_index = index >> 3;
+      ut_ad(byte_index == index / 8);
+
+      /* MAP array position */
+      m_map_index = byte_index >> 2;
+      ut_ad(m_map_index == byte_index / 4);
+
+      /* BIT position */
+      auto bit_pos = index & 31;
+      ut_ad(bit_pos == index % 32);
+
+      m_bit_mask = 1 << bit_pos;
+    }
+
+    /** Check value at specified index in BITMAP
+    @return true if the BIT is set */
+    operator bool() const {
+      auto &val = m_bitmap_ref[m_map_index];
+
+      if ((val & m_bit_mask) == 0) {
+        return (false);
+      }
+
+      return (true);
+    }
```

Comment:
This is a rather complex way of writing
```c++
operator bool() const { return m_bitmap_ref[m_map_index] & m_bit_mask; }
```
The entire `Bitmap_Operator_Impl` seems to a potential pessimization. I didn’t check the generated code yet. But, I think that we should consider using the existing `template<uint> class Bitmap`, instead of introducing a reimplementation of something similar.